extensions/modules/file/src/main/java/org/exist/xquery/modules/file/Sync.java

```diff
@@ -274,28 +454,124 @@ private void processRepoDesc(final Path targetFile, final DocumentImpl doc, fina
         }
     }
 
-	private void saveBinary(final Path targetDir, final BinaryDocument binary, final MemTreeBuilder output) {
-		final Path targetFile = targetDir.resolve(binary.getFileURI().toASCIIString());
-		try {
-			if (Files.exists(targetFile) && Files.getLastModifiedTime(targetFile).compareTo(FileTime.fromMillis(binary.getLastModified())) >= 0) {
-				return;
-			}
-
-			output.startElement(new QName("update", FileModule.NAMESPACE_URI), null);
-			output.addAttribute(new QName("file"), targetFile.toAbsolutePath().toString());
-			output.addAttribute(new QName("name"), binary.getFileURI().toString());
-			output.addAttribute(new QName("collection"), binary.getCollection().getURI().toString());
-			output.addAttribute(new QName("type"), "binary");
-			output.addAttribute(new QName("modified"), new DateTimeValue(new Date(binary.getLastModified())).getStringValue());
+    private void saveBinary(final Path targetFile, final BinaryDocument binary, final MemTreeBuilder output) {
+        try (final InputStream is = context.getBroker().getBinaryResource(binary)) {
+            Files.copy(is, targetFile, StandardCopyOption.REPLACE_EXISTING);
+        } catch (final Exception e) {
+            reportError(output, e.getMessage());
+        }
+    }
+
+    private void reportError(final MemTreeBuilder output, final String msg) {
+        output.startElement(Sync.FileErrorElement, null);
+        output.characters(msg);
+        output.endElement();
+    }
+
+    /**
+     * We need to convert to a relative path in relation to rootTargetAbsPath,
+     * as all the exclusion patterns are relative to rootTargetAbsPath.
+     *
+     * @param rootTargetAbsPath the root target (abs)path
+     * @param path              (abs)path to check for being excluded. Should be subdir of rootTargetAbsPath
+     * @param excludes          exclude patterns (in the convention of DirectoryScanner.match)
+     * @return true if the (rel)path in question is matched by some of the exclusion patterns
+     */
+    private static boolean isExcludedPath(String rootTargetAbsPath, Path path, List<String> excludes) {
+        if (excludes.isEmpty()) {
+            return false;
+        }
+        // root folder cannot be excluded
+        // path will then also be one character shorter than rootTargetApsPath
+        // and throw when attempting to construct the relative path
+        if (rootTargetAbsPath.startsWith(path.toString())) {
+            return false;
+        }
+
+        final String absPath = path.toAbsolutePath().toString();
+        final String relPath = absPath.substring(rootTargetAbsPath.length());
```

Comment:
There are functions in Java's `Path` class for getting relative paths, maybe better to use those.